web-console/src/utils/ingestion-spec.tsx

```diff
@@ -2181,17 +2181,41 @@ export function getPartitionRelatedTuningSpecFormFields(
           info: <>Total number of rows in segments waiting for being pushed.</>,
         },
         // partitionsSpec type: hashed
+        {
+          name: 'partitionsSpec.targetRowsPerSegment',
+          label: 'Target rows per segment',
+          type: 'number',
+          defined: (t: TuningConfig) =>
+            deepGet(t, 'partitionsSpec.type') === 'hashed' &&
+            !deepGet(t, 'partitionsSpec.numShards'),
+          info: (
+            <p>
+              If you have skewed data, you may want to set this field to generate evenly sized
+              segments.
+              <br />
+              <br /> A target row count for each partition. Each partition will have a row count
+              close to the target assuming evenly distributed keys. Defaults to 5 million if
+              numShards is null.
+            </p>
+          ),
+        },
         {
           name: 'partitionsSpec.numShards',
           label: 'Num shards',
           type: 'number',
-          defined: (t: TuningConfig) => deepGet(t, 'partitionsSpec.type') === 'hashed',
+          defined: (t: TuningConfig) =>
+            deepGet(t, 'partitionsSpec.type') === 'hashed' &&
+            !deepGet(t, 'partitionsSpec.targetRowsPerSegment'),
```

Comment:
Does this mean that either `numShards` should be set or `targetRowsPerSegment`? What is their relationship exactly? Does one have to be set? What if both are set? I am asking because it looks like there might need some validation logic?

Reply:
That's right, at most one is needed. If both are set, the backend will reject the spec. I forgot to add the validation logic. Will add in the next patch